### ISSUE TITLE
get_tightbbox doesn't account for markeredgewidth

### ISSUE THREAD

**Issue by brunobeltran** [matplotlib/matplotlib#16606]
Simple to solve problem, making issue just to refer to it in the PR.

### Bug report

For lines, `Axes.get_tightbbox` relies on `Line2D.get_window_extent`, which computes extra padding due to markers as 

```python
ms = (self._markersize / 72.0 * self.figure.dpi) * 0.5
```
but this can cut off half of the marker edge.

This phenomena which is only visible if you have a line that extends outside of a plot, and has `markeredgewidth > 1`. For simple markers, (like 'S', 'O'), the code should instead read

```python
extra_pts = self._markersize + self._markeredgewidth
ms = (extra_pts / 72.0 * self.figure.dpi) * 0.5
```
I'll edit this shortly to include code that reproduces the bug, after I document for all the common markers what the expected behavior should be.

**Matplotlib version**

  * Operating system: Linux
  * Matplotlib version: 3.2.0rc2.post1594.dev0+gead45c7e6
  * Matplotlib backend: module://ipykernel.pylab.backend_inline`
  * Python version: 3.7.3
MPL installed from Github, Python is miniconda.

**Comment by brunobeltran** [comment]
This seems like a pretty niche use case, and begs the question what I'm doing to make it necessary:

Say you're making a figure for publication, so that the font size, and figure width is fixed. In modern biology, it's not uncommon to be VERY short on figure space, leading to figures with funny sizes, like the following:

```python
fig, axs = plt.subplots(2, figsize=(2,4))
for ax in axs:
    for dash, marker in zip(['-', '-.', '--'], ['D', 'd', 'o']):
        N = 5
        x = np.random.rand(N)
        y = np.random.rand(N)
        ax.plot(x, y, 
                 marker=marker, markersize=12, markeredgewidth=5,
                 fillstyle='full', markerfacecolor='white', 
                 linestyle=dash, lw=3, clip_on=False)
```

![example_overlap](https://user-images.githubusercontent.com/1475390/76028337-4073ab00-5ee7-11ea-8b42-b8b0b681ddb6.png)

As you can see, the markers overlap strongly across plots. I can easily fix this by just spacing the axes myself. But for that I need to know the `ax.get_tightbbox`, which is currently wrong. 

Normally being a little bit off wouldn't be a big deal, but in these figures where the marker edge width is comparable to the figure size, (say >1% of total figure size or so) then being off by a bit can make some of the axes look noticeably misspaced.

I like to think I understand the internals of matplotlib more than the average use, and even when being pretty clever, I simply don't have a way to space the axes nicely without knowing where the bbox's are

```python
from bruno_util.plotting import make_at_aspect
def random_large_markers(ax, N=5):
    for dash, marker in zip(['-', '-.', '--'], ['D', 'd', 'o']):
        ax.plot(np.random.rand(N), np.random.rand(N), 
                marker=marker, markersize=12, markeredgewidth=5,
                fillstyle='full', markerfacecolor='white', 
                linestyle=dash, lw=3, clip_on=False)
    ax.set_xlim([0, 1])
    ax.set_ylim([0, 1])
    ax.set_xlabel('x axis')
    ax.set_ylabel('y axis')

golden_ratio = (1 + np.sqrt(5))/2
col_width = 2 # inches. imposed by journal format
fig = make_at_aspect([random_large_markers, random_large_markers], 
                     [1/golden_ratio, 1/golden_ratio], col_width)
```
![example_overlap_stack](https://user-images.githubusercontent.com/1475390/76031967-afec9900-5eed-11ea-92de-876aace2d9f9.png)

**Comment by brunobeltran** [comment]
Originally I had hoped to just add `markeredgewidth/2` padding to the existing bbox calculation code, but since this underestimates the extents of corners, if I'm not careful I can cause my markers to be cut off (which looks V ugly). 

![test](https://user-images.githubusercontent.com/1475390/76033563-600fd100-5ef1-11ea-9c8b-9817e049e37c.png)

**Comment by brunobeltran** [comment]
I'm sure you already know this, but that would be a large overestimate for some markers (e.g. 'o', 's', etc.), where `markeredgewidth/2` is the correct amount. 

So using the code in my PR would allow the figure boundary to be set "exactly flush" with the markers, instead of leaving annoying whitespace that has to be removed in Illustrator.

But if it is agreed that an overestimate is what we want (even though I already have the code to compute the bbox exactly written), then I can compute what the correct amount to add is (given the miter limit). I'd guess that it will be a hair larger than `+markeredgewidth`, because of some sharp corners like the top of `marker='d'`, but you might be right, I'd have to check.

**Comment by anntzer** [comment]
Yes, you would lose a couple of pixels for circles.  I think if your figure is so cramped that you even need those (I've played the game of 6 figures x 25 subfigures each too...) then perhaps you can either slightly change the figure design, or even(!) manually set slightly larger margins :)

**Comment by brunobeltran** [comment]
At that point the fix is more about matplotlib "feeling professional" (in the sense that it works exactly and correctly as advertised) vs it being a real problem (I can always just trip some extra whitespace in illustrator....

**Comment by anntzer** [comment]
I got your point :-) let's keep the discussion on the PR side to avoid splitting it in two places.

**Comment by anntzer** [comment]
Let's keep this open, if open to keep track of #17198 (which seems useful in itself).
